pkg/cloud/services/ec2/instances.go

```diff
@@ -321,43 +321,43 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 	case failureDomain != nil:
 		if scope.AWSMachine.Spec.PublicIP != nil && *scope.AWSMachine.Spec.PublicIP {
 			subnets := s.scope.Subnets().FilterPublic().FilterByZone(*failureDomain)
-			if len(subnets) == 0 {
-				errMessage := fmt.Sprintf("failed to run machine %q with public IP, no public subnets available in availability zone %q",
-					scope.Name(), *failureDomain)
-				record.Warnf(scope.AWSMachine, "FailedCreate", errMessage)
-				return "", awserrors.NewFailedDependency(errMessage)
+			if subnetID := getSubnetID(subnets); subnetID != "" {
+				return subnetID, nil
 			}
-			return subnets[0].ID, nil
-		}
-
-		subnets := s.scope.Subnets().FilterPrivate().FilterByZone(*failureDomain)
-		if len(subnets) == 0 {
-			errMessage := fmt.Sprintf("failed to run machine %q, no subnets available in availability zone %q",
+			errMessage := fmt.Sprintf("failed to run machine %q with public IP, no public subnets available in availability zone %q",
```

Comment:
Using same log for having no public subnets in the list vs having public subnets but missing their subnet IDs is misleading. Might worth keeping `if len(subnets) == 0` check.